### ISSUE TITLE
Projection pre and post order

### ISSUE THREAD

**Issue by timotheeguerin** [microsoft/typespec#1500]
Right now we have issues with projection removing items but still having decorators run on them or reference those items.

## Proposal

1. Define 2 stage for projection to run `pre` and `post`
2. Allow to define projection for member elements

Currently the flow for a projection is:

1. Start projecting type
2. Recursively project children
3. Run decorators
4. Run the projection (Here we might remove the type and now decorators have already been run on this type and its children)

With the proposed changed we insert a new step:

1. Start projecting type
2. Run `pre` projection (Here we would decide to remove the type and so we don't run the decorator nor have the children created.)
3. Recursively project children
4. Run decorators
5. Run `post` projection

## Syntax change

```cadl
projection model#v {
  to {
    // This remains the `post` order projection
  }

  pre to {
    // Here define the step that run pre.
  }

  // Alternative name
  preTo {

  }
}
```

## Target member elements

Use reflection types?

```cadl
projection Cadl.Reflection.ModelProperty#v {

}
```

**Comment by markcowl** [comment]
- selector keywords
  - names? (what are they)
  - name clashes with identifiers in specs (global::<identifier>)
- metadata types from Cadl.Reflection

**Comment by tjprescott** [comment]
Related issue: https://github.com/Azure/cadl-azure/issues/2368

**Comment by markcowl** [comment]
est: 13

**Comment by allenjzhang** [comment]
instead of `preTo` or `pre To`, thoughts on `prepare`?

**Comment by bterlson** [comment]
@allenjzhang do you mean `prepareTo`? Note that we would also need `prepareFrom`. I am weakly in favor of `pre` terminology since we're talking about a preorder and postorder traversal, but could be convinced otherwise.

**Comment by nguerrera** [comment]
> There was a question being asked on the precedence over @friendlyName

friendlyName doesn't really have any precedence. It's data on the side that you can query and choose to use or not. However, I think the friendlyName would remain intact in the face of projection unless the projection applied a new friendlyName decorator. I'm going to consider this a completely separate discussion because I think it applies with or without "pre" projections.

**Comment by nguerrera** [comment]
Since this is already an implementation issue in the current sprint, I will go ahead and use `pre to` and `pre from` since subjectively that's my favorite of the options I've seen so far, but changing this syntax will be the smallest part of the change and I will happily change it if another consensus emerges.

**Comment by nguerrera** [comment]
Changed my mind, it's easier to use preTo and preFrom in the implementation right now. So I'll save my energy until we're ready to make a call and stage the feature using preTo and preFrom at least for now.

**Comment by bterlson** [comment]
I wouldn't object to `preTo` and `preFrom`, though I think the space form is also pretty easy (parse two contextual keywords via parseIdentifier() instead of one?)

**Comment by nguerrera** [comment]
yeah, it's easy to parse, but it's a little easier to have an single thing that must be unique than two. It's a minor thing, but it annoyed me just enough to not want to bother while I haven't gotten to the actual hard problems yet.

**Comment by nguerrera** [comment]
Not sure if this was already covered in design discussion, but I'm not finding a record of it, so putting some questions here as I finally start to dig in.

What does "pre" projection see for children? No children yet or unprojected children? Can pre projection mutate/add/remove children? 

Depending on the answers here, it seems we might need to do a deep no-op clone before we can run pre-projections? We seem to rely on projecting children to copy them before projection runs now.

Is the main use case identified to delete a type and short-circuit children projection? Would it be simpler to have a more specialized callback that is not implied to be as capable as to/from but rather just a chance to return delete-or-not with read-only access to unprojected type? A sort of projection filter callback?

**Comment by nguerrera** [comment]
> Allow to define projection for member elements

I missed this part of the proposal until now... It's not immediately obvious why this is tied to this.

**Comment by nguerrera** [comment]
Ah! Because currently members are projected in a foreach in the .tsp projection code for their parent. This would not be able to take advantage of the early deletion for members if we don't also add that. 

Using the reflection types for that is a little weird since it already would have the meaning of projecting the reflection types itself. Could also allow additional contextual keywords there. 

We could revive the names we were going to use for decorator signatures:

`projection modelproperty#v`
`projection unionvariant#v`
`projection enummember#v`

These didn't work for decorator signatures since they need to be able to appear on rhs of `:` anywhere in the signature (not just at top-level) everywhere and therefore couldn't be (very) contextual. But here, it's easy I think to allow them contextually in just this position. I'll try that.

I'm putting together a draft here, but this will need a design discussion to go over this syntax and some of the issues I raised earlier about mutation in pre-projection and such.

**Comment by nguerrera** [comment]
Interface members are a little special since they are operations and we already have `projection op`. It looks like this is invoked already on interface members if I'm reading right. If so, I guess we can move code from `forEach` in `projection interface` to `projection op` and get away with only the three new cases for model,union, and enum members.

**Comment by nguerrera** [comment]
Draft is up at #1673, will file a separate issue for remaining design questions.

**Comment by nguerrera** [comment]
#1674 has the remaining design questions.
